File: LiveDubV1/app/src/main/java/com/vinh/livedub/LiveDubService.java
package com.vinh.livedub;

import android.app.*;
import android.content.*;
import android.media.*;
import android.media.projection.MediaProjection;
import android.media.projection.MediaProjectionManager;
import android.os.*;
import android.speech.*;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;

import com.google.mlkit.common.model.DownloadConditions;
import com.google.mlkit.nl.translate.*;

import java.io.FileOutputStream;
import java.io.OutputStream;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

public class LiveDubService extends Service {
    private static final String CHANNEL_ID = "livedub_channel";
    private static final int NOTIF_ID = 1001;

    private MediaProjection projection;
    private AudioRecord audioRecord;
    private Thread captureThread;
    private volatile boolean running = false;

    private SpeechRecognizer recognizer;
    private ParcelFileDescriptor[] speechPipe;
    private volatile OutputStream speechOut;
    private final Handler main = new Handler(Looper.getMainLooper());

    private Translator translator;
    private volatile boolean translatorReady = false;
    private TextToSpeech tts;
    private volatile boolean ttsReady = false;
    private AudioManager audioManager;
    private AudioFocusRequest focusRequest;

    private String lastPartialBase = "";
    private String lastHandledText = "";
    private long lastPartialAt = 0;
    private final AtomicLong seq = new AtomicLong();
    private volatile long lastSpokenSeq = 0;

    @Override
    public void onCreate() {
        super.onCreate();
        audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
        createChannel();
        initTranslator();
        initTts();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        startForeground(NOTIF_ID, buildNotification("Đang khởi động…"));
        if (intent == null) return START_NOT_STICKY;

        int resultCode = intent.getIntExtra("resultCode", Activity.RESULT_CANCELED);
        Intent data = intent.getParcelableExtra("resultData");
        if (resultCode != Activity.RESULT_OK || data == null) {
            updateNotification("Thiếu quyền capture");
            stopSelf();
            return START_NOT_STICKY;
        }

        try {
            MediaProjectionManager mpm = (MediaProjectionManager) getSystemService(MEDIA_PROJECTION_SERVICE);
            projection = mpm.getMediaProjection(resultCode, data);
            projection.registerCallback(new MediaProjection.Callback() {
                @Override public void onStop() { stopSelf(); }
            }, main);
            startPipeline();
        } catch (Throwable t) {
            updateNotification("Lỗi khởi tạo: " + shortMsg(t));
            stopSelf();
        }
        return START_NOT_STICKY;
    }

    private void startPipeline() throws Exception {
        AudioPlaybackCaptureConfiguration config = new AudioPlaybackCaptureConfiguration.Builder(projection)
                .addMatchingUsage(AudioAttributes.USAGE_MEDIA)
                .excludeUid(android.os.Process.myUid())
                .build();

        AudioFormat format = new AudioFormat.Builder()
                .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                .setSampleRate(16000)
                .setChannelMask(AudioFormat.CHANNEL_IN_MONO)
                .build();

        int min = AudioRecord.getMinBufferSize(16000, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);
        int bufferBytes = Math.max(min * 4, 32768);

        audioRecord = new AudioRecord.Builder()
                .setAudioFormat(format)
                .setBufferSizeInBytes(bufferBytes)
                .setAudioPlaybackCaptureConfig(config)
                .build();

        if (audioRecord.getState() != AudioRecord.STATE_INITIALIZED) {
            throw new IllegalStateException("AudioRecord không khởi tạo được");
        }

        setupRecognizer();
        audioRecord.startRecording();
        running = true;
        captureThread = new Thread(this::captureLoop, "LiveDubCapture");
        captureThread.start();
        updateNotification("Đang nghe audio nội bộ • Anh → Việt");
    }

    private void captureLoop() {
        byte[] buffer = new byte[4096];
        while (running) {
            try {
                int n = audioRecord.read(buffer, 0, buffer.length, AudioRecord.READ_BLOCKING);
                OutputStream out = speechOut;
                if (n > 0 && out != null) {
                    out.write(buffer, 0, n);
                    out.flush();
                }
            } catch (Throwable e) {
                if (running) main.post(() -> restartRecognizer("Luồng nhận dạng được nối lại"));
                try { Thread.sleep(250); } catch (InterruptedException ignored) {}
            }
        }
    }

    private void setupRecognizer() {
        main.post(() -> {
            try {
                if (!SpeechRecognizer.isRecognitionAvailable(this)) {
                    updateNotification("Máy không có dịch vụ nhận dạng giọng nói");
                    return;
                }
                recognizer = SpeechRecognizer.createSpeechRecognizer(this);
                recognizer.setRecognitionListener(new RecognitionListener() {
                    @Override public void onReadyForSpeech(Bundle params) { updateNotification("Live Sync đang chạy • mở YouTube/X"); }
                    @Override public void onBeginningOfSpeech() {}
                    @Override public void onRmsChanged(float rmsdB) {}
                    @Override public void onBufferReceived(byte[] buffer) {}
                    @Override public void onEndOfSpeech() {}
                    @Override public void onError(int error) {
                        if (running) main.postDelayed(() -> restartRecognizer("Nhận dạng lỗi " + error + ", đang nối lại"), 600);
                    }
                    @Override public void onResults(Bundle results) { handleBundle(results, true); }
                    @Override public void onPartialResults(Bundle partialResults) { handleBundle(partialResults, false); }
                    @Override public void onEvent(int eventType, Bundle params) {}
                    @Override public void onSegmentResults(Bundle segmentResults) { handleBundle(segmentResults, true); }
                    @Override public void onEndOfSegmentedSession() {
                        if (running) main.postDelayed(() -> restartRecognizer("Bắt đầu phiên nhận dạng mới"), 250);
                    }
                });
                startRecognitionSession();
            } catch (Throwable t) {
                updateNotification("Không mở được nhận dạng: " + shortMsg(t));
            }
        });
    }

    private void startRecognitionSession() throws Exception {
        closeSpeechPipe();
        speechPipe = ParcelFileDescriptor.createPipe();
        speechOut = new FileOutputStream(speechPipe[1].getFileDescriptor());

        Intent r = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        r.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        r.putExtra(RecognizerIntent.EXTRA_LANGUAGE, "en-US");
        r.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true);
        r.putExtra(RecognizerIntent.EXTRA_AUDIO_SOURCE, speechPipe[0]);
        r.putExtra(RecognizerIntent.EXTRA_AUDIO_SOURCE_CHANNEL_COUNT, 1);
        r.putExtra(RecognizerIntent.EXTRA_AUDIO_SOURCE_ENCODING, AudioFormat.ENCODING_PCM_16BIT);
        r.putExtra(RecognizerIntent.EXTRA_AUDIO_SOURCE_SAMPLING_RATE, 16000);
        r.putExtra(RecognizerIntent.EXTRA_SEGMENTED_SESSION, RecognizerIntent.EXTRA_AUDIO_SOURCE);
        recognizer.startListening(r);
    }

    private void restartRecognizer(String status) {
        updateNotification(status);
        try { if (recognizer != null) recognizer.cancel(); } catch (Throwable ignored) {}
        try { startRecognitionSession(); } catch (Throwable t) { updateNotification("Không nối lại được nhận dạng: " + shortMsg(t)); }
    }

    private void handleBundle(Bundle b, boolean isFinal) {
        if (b == null) return;
        ArrayList<String> list = b.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
        if (list == null || list.isEmpty()) return;
        String text = list.get(0) == null ? "" : list.get(0).trim();
        if (text.length() < 2) return;

        if (!isFinal) {
            long now = SystemClock.elapsedRealtime();
            if (now - lastPartialAt < 1100 || text.length() < 12) return;
            lastPartialAt = now;
            String delta = suffixDelta(lastPartialBase, text);
            if (delta.split("\\s+").length < 3) return;
            lastPartialBase = text;
            processText(delta);
        } else {
            String delta = suffixDelta(lastPartialBase, text);
            lastPartialBase = "";
            if (delta.length() >= 2) processText(delta);
        }
    }

    private String suffixDelta(String oldText, String newText) {
        if (oldText == null || oldText.isEmpty()) return newText;
        if (newText.equalsIgnoreCase(oldText)) return "";
        String[] a = oldText.split("\\s+");
        String[] b = newText.split("\\s+");
        int i = 0;
        while (i < a.length && i < b.length && a[i].equalsIgnoreCase(b[i])) i++;
        if (i >= b.length) return "";
        StringBuilder sb = new StringBuilder();
        for (int j = i; j < b.length; j++) { if (sb.length() > 0) sb.append(' '); sb.append(b[j]); }
        return sb.toString();
    }

    private void processText(String source) {
        source = source.trim();
        if (source.isEmpty() || source.equalsIgnoreCase(lastHandledText)) return;
        lastHandledText = source;
        if (!translatorReady) { updateNotification("Đang tải mô hình dịch Anh → Việt…"); return; }

        final long id = seq.incrementAndGet();
        translator.translate(source)
                .addOnSuccessListener(vi -> {
                    if (vi == null || vi.trim().isEmpty() || id < lastSpokenSeq) return;
                    lastSpokenSeq = id;
                    speakLive(vi.trim(), id);
                })
                .addOnFailureListener(e -> updateNotification("Lỗi dịch: " + shortMsg(e)));
    }

    private void speakLive(String vi, long id) {
        updateNotification("VI: " + vi);
        if (!ttsReady) return;
        boolean chasing = tts.isSpeaking();
        tts.setSpeechRate(chasing ? 1.28f : 1.10f);
        requestDuck();
        Bundle p = new Bundle();
        p.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, 1.0f);
        tts.speak(vi, chasing ? TextToSpeech.QUEUE_FLUSH : TextToSpeech.QUEUE_ADD, p, "dub-" + id);
    }

    private void initTranslator() {
        TranslatorOptions options = new TranslatorOptions.Builder()
                .setSourceLanguage(TranslateLanguage.ENGLISH)
                .setTargetLanguage(TranslateLanguage.VIETNAMESE)
                .build();
        translator = Translation.getClient(options);
        translator.downloadModelIfNeeded(new DownloadConditions.Builder().build())
                .addOnSuccessListener(v -> { translatorReady = true; updateNotification("Sẵn sàng • Anh → Việt"); })
                .addOnFailureListener(e -> updateNotification("Không tải được mô hình dịch: " + shortMsg(e)));
    }

    private void initTts() {
        tts = new TextToSpeech(this, status -> {
            if (status == TextToSpeech.SUCCESS) {
                tts.setLanguage(Locale.forLanguageTag("vi-VN"));
                tts.setSpeechRate(1.10f);
                tts.setAudioAttributes(new AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ASSISTANCE_ACCESSIBILITY)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH).build());
                tts.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                    @Override public void onStart(String utteranceId) {}
                    @Override public void onDone(String utteranceId) { abandonDuck(); }
                    @Override public void onError(String utteranceId) { abandonDuck(); }
                });
                ttsReady = true;
            }
        });
    }

    private void requestDuck() {
        try {
            if (Build.VERSION.SDK_INT >= 26) {
                focusRequest = new AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
                        .setAudioAttributes(new AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_ASSISTANCE_ACCESSIBILITY).setContentType(AudioAttributes.CONTENT_TYPE_SPEECH).build())
                        .setOnAudioFocusChangeListener(change -> {})
                        .build();
                audioManager.requestAudioFocus(focusRequest);
            } else {
                audioManager.requestAudioFocus(null, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK);
            }
        } catch (Throwable ignored) {}
    }

    private void abandonDuck() {
        try {
            if (Build.VERSION.SDK_INT >= 26 && focusRequest != null) audioManager.abandonAudioFocusRequest(focusRequest);
            else audioManager.abandonAudioFocus(null);
        } catch (Throwable ignored) {}
    }

    private void createChannel() {
        if (Build.VERSION.SDK_INT >= 26) {
            NotificationChannel c = new NotificationChannel(CHANNEL_ID, "LiveDub Việt", NotificationManager.IMPORTANCE_LOW);
            c.setDescription("Trạng thái lồng tiếng thời gian thực");
            ((NotificationManager) getSystemService(NOTIFICATION_SERVICE)).createNotificationChannel(c);
        }
    }

    private Notification buildNotification(String text) {
        Intent open = new Intent(this, MainActivity.class);
        PendingIntent pi = PendingIntent.getActivity(this, 0, open, PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
        return new Notification.Builder(this, CHANNEL_ID)
                .setContentTitle("LiveDub Việt • Live Sync")
                .setContentText(text)
                .setSmallIcon(android.R.drawable.ic_btn_speak_now)
                .setOngoing(true)
                .setContentIntent(pi)
                .build();
    }

    private void updateNotification(String text) {
        try { ((NotificationManager) getSystemService(NOTIFICATION_SERVICE)).notify(NOTIF_ID, buildNotification(text)); } catch (Throwable ignored) {}
    }

    private String shortMsg(Throwable t) {
        String s = t.getMessage();
        if (s == null || s.trim().isEmpty()) s = t.getClass().getSimpleName();
        return s.length() > 70 ? s.substring(0, 70) : s;
    }

    private void closeSpeechPipe() {
        try { if (speechOut != null) speechOut.close(); } catch (Throwable ignored) {}
        speechOut = null;
        if (speechPipe != null) {
            try { speechPipe[0].close(); } catch (Throwable ignored) {}
            try { speechPipe[1].close(); } catch (Throwable ignored) {}
        }
        speechPipe = null;
    }

    @Override
    public void onDestroy() {
        running = false;
        try { if (audioRecord != null) { audioRecord.stop(); audioRecord.release(); } } catch (Throwable ignored) {}
        try { if (recognizer != null) { recognizer.cancel(); recognizer.destroy(); } } catch (Throwable ignored) {}
        closeSpeechPipe();
        try { if (projection != null) projection.stop(); } catch (Throwable ignored) {}
        try { if (translator != null) translator.close(); } catch (Throwable ignored) {}
        try { if (tts != null) { tts.stop(); tts.shutdown(); } } catch (Throwable ignored) {}
        abandonDuck();
        super.onDestroy();
    }

    @Override public IBinder onBind(Intent intent) { return null; }
}

File: LiveDubVI/app/src/main/java/com/vinh/livedubvi/LiveDubService.kt
package com.vinh.livedubvi

import android.app.*
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.media.*
import android.media.projection.MediaProjection
import android.media.projection.MediaProjectionManager
import android.os.Build
import android.os.IBinder
import android.os.Process
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import androidx.core.app.NotificationCompat
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.Translator
import com.google.mlkit.nl.translate.TranslatorOptions
import dev.ffmpegkit.whisper.Whisper
import dev.ffmpegkit.whisper.WhisperConfig
import dev.ffmpegkit.whisper.WhisperModel
import kotlinx.coroutines.*
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.util.Locale
import java.util.UUID
import java.util.concurrent.atomic.AtomicReference
import kotlin.coroutines.resume
import kotlin.math.abs

class LiveDubService : Service(), TextToSpeech.OnInitListener {
    companion object {
        const val ACTION_START = "com.vinh.livedubvi.START"
        const val ACTION_STOP = "com.vinh.livedubvi.STOP"
        const val EXTRA_RESULT_CODE = "resultCode"
        const val EXTRA_RESULT_DATA = "resultData"
        const val CHANNEL_ID = "livedub_capture"
        const val NOTIFICATION_ID = 7186

        @Volatile var lastStatus: String = "Sẵn sàng."
        @Volatile var running: Boolean = false
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var projection: MediaProjection? = null
    private var audioRecord: AudioRecord? = null
    private var whisperModel: WhisperModel? = null
    private var translator: Translator? = null
    private var tts: TextToSpeech? = null
    private var ttsReady = false
    private val latestChunk = AtomicReference<File?>(null)
    private val pendingSpeech = AtomicReference<String?>(null)
    private var lastOriginal = ""
    private lateinit var audioManager: AudioManager
    private var focusRequest: AudioFocusRequest? = null

    override fun onCreate() {
        super.onCreate()
        audioManager = getSystemService(Context.AUDIO_SERVICE) as AudioManager
        createChannel()
        tts = TextToSpeech(this, this)
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_STOP -> stopEverything()
            ACTION_START -> {
                if (running) return START_STICKY
                startForegroundNow()
                val resultCode = intent.getIntExtra(EXTRA_RESULT_CODE, Activity.RESULT_CANCELED)
                val resultData = if (Build.VERSION.SDK_INT >= 33) {
                    intent.getParcelableExtra(EXTRA_RESULT_DATA, Intent::class.java)
                } else {
                    @Suppress("DEPRECATION") intent.getParcelableExtra(EXTRA_RESULT_DATA)
                }
                if (resultCode != Activity.RESULT_OK || resultData == null) {
                    lastStatus = "Không nhận được quyền MediaProjection."
                    stopEverything()
                } else {
                    running = true
                    scope.launch { startPipeline(resultCode, resultData) }
                }
            }
        }
        return START_STICKY
    }

    private fun startForegroundNow() {
        val stopIntent = Intent(this, LiveDubService::class.java).apply { action = ACTION_STOP }
        val stopPending = PendingIntent.getService(
            this, 2, stopIntent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_btn_speak_now)
            .setContentTitle("LiveDub Việt")
            .setContentText("Đang chuẩn bị lồng tiếng từ âm thanh nội bộ")
            .setOngoing(true)
            .addAction(android.R.drawable.ic_menu_close_clear_cancel, "Dừng", stopPending)
            .build()
        if (Build.VERSION.SDK_INT >= 29) {
            startForeground(NOTIFICATION_ID, notification, ServiceInfo.FOREGROUND_SERVICE_TYPE_MEDIA_PROJECTION)
        } else {
            startForeground(NOTIFICATION_ID, notification)
        }
    }

    private suspend fun startPipeline(resultCode: Int, resultData: Intent) {
        try {
            lastStatus = "Đang nạp Whisper tiny.en…"
            whisperModel = Whisper.loadModelFromAsset(this, "models/ggml-tiny.en.bin")
            lastStatus = "Đang chuẩn bị mô hình dịch Anh → Việt…"
            translator = Translation.getClient(
                TranslatorOptions.Builder()
                    .setSourceLanguage(TranslateLanguage.ENGLISH)
                    .setTargetLanguage(TranslateLanguage.VIETNAMESE)
                    .build()
            )
            awaitTranslatorModel(translator!!)

            val manager = getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
            projection = manager.getMediaProjection(resultCode, resultData)
            projection?.registerCallback(object : MediaProjection.Callback() {
                override fun onStop() {
                    lastStatus = "Android đã dừng phiên bắt âm thanh."
                    stopEverything()
                }
            }, null)

            startAudioCapture()
            scope.launch { transcriptionWorker() }
            lastStatus = "Đang nghe âm thanh nội bộ • EN → VI • Live Sync"
        } catch (t: Throwable) {
            lastStatus = "Lỗi khởi động: ${t.message ?: t.javaClass.simpleName}"
            stopEverything()
        }
    }

    private fun startAudioCapture() {
        val p = projection ?: error("MediaProjection chưa sẵn sàng")
        val captureConfig = AudioPlaybackCaptureConfiguration.Builder(p)
            .addMatchingUsage(AudioAttributes.USAGE_MEDIA)
            .addMatchingUsage(AudioAttributes.USAGE_GAME)
            .addMatchingUsage(AudioAttributes.USAGE_UNKNOWN)
            .excludeUid(Process.myUid())
            .build()

        val format = AudioFormat.Builder()
            .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
            .setSampleRate(16000)
            .setChannelMask(AudioFormat.CHANNEL_IN_MONO)
            .build()

        val min = AudioRecord.getMinBufferSize(16000, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT)
        audioRecord = AudioRecord.Builder()
            .setAudioFormat(format)
            .setBufferSizeInBytes(maxOf(min * 4, 32768))
            .setAudioPlaybackCaptureConfig(captureConfig)
            .build()

        audioRecord?.startRecording()
        scope.launch(Dispatchers.IO) { captureLoop() }
    }

    private suspend fun captureLoop() {
        val record = audioRecord ?: return
        val readBuf = ByteArray(4096)
        val chunk = ByteArrayOutputStream(96000)
        var voicedSamples = 0L
        val targetBytes = 16000 * 2 * 24 / 10
        val overlapBytes = 16000 * 2 * 35 / 100
        var carry = ByteArray(0)

        while (running && record.recordingState == AudioRecord.RECORDSTATE_RECORDING) {
            val n = record.read(readBuf, 0, readBuf.size, AudioRecord.READ_BLOCKING)
            if (n <= 0) continue
            if (chunk.size() == 0 && carry.isNotEmpty()) chunk.write(carry)
            chunk.write(readBuf, 0, n)
            voicedSamples += countVoiced(readBuf, n)

            if (chunk.size() >= targetBytes) {
                val pcm = chunk.toByteArray()
                val voiceRatio = voicedSamples.toDouble() / (pcm.size / 2).coerceAtLeast(1)
                if (voiceRatio > 0.035) {
                    val file = File(cacheDir, "cap_${System.nanoTime()}.wav")
                    writeWav(file, pcm, 16000, 1)
                    submitLatest(file)
                    lastStatus = "Đã bắt âm thanh • đang nhận dạng…"
                } else {
                    lastStatus = "Đang nghe… chưa thấy giọng nói rõ"
                }
                carry = pcm.copyOfRange((pcm.size - overlapBytes).coerceAtLeast(0), pcm.size)
                chunk.reset()
                voicedSamples = 0
            }
        }
    }

    private fun countVoiced(bytes: ByteArray, length: Int): Long {
        var voiced = 0L
        var i = 0
        while (i + 1 < length) {
            val s = ((bytes[i + 1].toInt() shl 8) or (bytes[i].toInt() and 0xff)).toShort().toInt()
            if (abs(s) > 520) voiced++
            i += 2
        }
        return voiced
    }

    private fun submitLatest(file: File) {
        val old = latestChunk.getAndSet(file)
        if (old != null && old != file) old.delete()
    }

    private suspend fun transcriptionWorker() {
        while (running) {
            val file = latestChunk.getAndSet(null)
            if (file == null) {
                delay(80)
                continue
            }
            try {
                val model = whisperModel ?: continue
                val started = System.currentTimeMillis()
                val result = Whisper.transcribe(
                    model,
                    file.absolutePath,
                    WhisperConfig(language = "en", threads = 6, printTimestamps = false)
                )
                val raw = cleanWhisper(result.text)
                val fresh = dedupe(lastOriginal, raw)
                if (raw.isNotBlank()) lastOriginal = raw
                if (fresh.isBlank() || isWhisperNoise(fresh)) continue

                lastStatus = "EN: ${fresh.take(90)}"
                val vi = translate(fresh)
                if (vi.isNotBlank()) {
                    val elapsed = System.currentTimeMillis() - started
                    lastStatus = "VI: ${vi.take(90)} • xử lý ${elapsed}ms"
                    speakLatest(vi, elapsed)
                }
            } catch (t: Throwable) {
                lastStatus = "Nhận dạng lỗi: ${t.message ?: t.javaClass.simpleName}"
            } finally {
                file.delete()
            }
        }
    }

    private fun cleanWhisper(s: String): String = s
        .replace(Regex("\\[[^]]*]"), " ")
        .replace(Regex("\\([^)]*(music|applause|laughter)[^)]*\\)", RegexOption.IGNORE_CASE), " ")
        .replace(Regex("\\s+"), " ")
        .trim()

    private fun isWhisperNoise(s: String): Boolean {
        val x = s.lowercase(Locale.US).trim().trim('.', '!', '?', ' ')
        return x.isBlank() || x in setOf("thank you", "thanks for watching", "you")
    }

    private fun dedupe(previous: String, current: String): String {
        if (current.isBlank() || previous.isBlank()) return current
        val p = previous.trim().split(Regex("\\s+"))
        val c = current.trim().split(Regex("\\s+"))
        val max = minOf(10, p.size, c.size)
        for (k in max downTo 2) {
            val a = p.takeLast(k).joinToString(" ").lowercase(Locale.US).trimPunct()
            val b = c.take(k).joinToString(" ").lowercase(Locale.US).trimPunct()
            if (a == b) return c.drop(k).joinToString(" ")
        }
        return current
    }

    private fun String.trimPunct(): String = replace(Regex("[^a-z0-9' ]"), "").replace(Regex("\\s+"), " ").trim()

    private suspend fun awaitTranslatorModel(t: Translator) = suspendCancellableCoroutine<Unit> { cont ->
        t.downloadModelIfNeeded(DownloadConditions.Builder().build())
            .addOnSuccessListener { if (cont.isActive) cont.resume(Unit) }
            .addOnFailureListener { if (cont.isActive) cont.resume(Unit) }
    }

    private suspend fun translate(text: String): String {
        val t = translator ?: return text
        return suspendCancellableCoroutine { cont ->
            t.translate(text)
                .addOnSuccessListener { out -> if (cont.isActive) cont.resume(out) }
                .addOnFailureListener { if (cont.isActive) cont.resume(text) }
        }
    }

    private fun speakLatest(text: String, processingMs: Long) {
        if (!ttsReady || text.isBlank()) return
        val engine = tts ?: return
        if (engine.isSpeaking) {
            pendingSpeech.getAndSet(text)
            return
        }
        val speed = when {
            processingMs > 2600 -> 1.32f
            processingMs > 1700 -> 1.24f
            else -> 1.16f
        }
        engine.setSpeechRate(speed)
        requestDuck()
        engine.speak(text, TextToSpeech.QUEUE_FLUSH, null, UUID.randomUUID().toString())
    }

    override fun onInit(status: Int) {
        if (status != TextToSpeech.SUCCESS) {
            ttsReady = false
            return
        }
        val engine = tts ?: return
        engine.language = Locale("vi", "VN")
        engine.setAudioAttributes(
            AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_ASSISTANCE_ACCESSIBILITY)
                .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                .build()
        )
        engine.setSpeechRate(1.16f)
        engine.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
            override fun onStart(utteranceId: String?) = Unit
            override fun onDone(utteranceId: String?) {
                abandonDuck()
                val next = pendingSpeech.getAndSet(null)
                if (!next.isNullOrBlank() && running) speakLatest(next, 2200)
            }
            @Deprecated("Deprecated in Java")
            override fun onError(utteranceId: String?) {
                abandonDuck()
            }
            override fun onError(utteranceId: String?, errorCode: Int) {
                abandonDuck()
            }
        })
        ttsReady = true
    }

    private fun requestDuck() {
        if (Build.VERSION.SDK_INT >= 26) {
            val req = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
                .setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_ASSISTANCE_ACCESSIBILITY)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .build()
                )
                .setOnAudioFocusChangeListener { }
                .build()
            focusRequest = req
            audioManager.requestAudioFocus(req)
        } else {
            @Suppress("DEPRECATION")
            audioManager.requestAudioFocus(null, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK)
        }
    }

    private fun abandonDuck() {
        if (Build.VERSION.SDK_INT >= 26) {
            focusRequest?.let { audioManager.abandonAudioFocusRequest(it) }
            focusRequest = null
        } else {
            @Suppress("DEPRECATION") audioManager.abandonAudioFocus(null)
        }
    }

    private fun writeWav(file: File, pcm: ByteArray, sampleRate: Int, channels: Int) {
        FileOutputStream(file).use { out ->
            val byteRate = sampleRate * channels * 2
            val dataSize = pcm.size
            val totalSize = dataSize + 36
            fun le32(v: Int) {
                out.write(v and 0xff); out.write((v shr 8) and 0xff); out.write((v shr 16) and 0xff); out.write((v shr 24) and 0xff)
            }
            fun le16(v: Int) { out.write(v and 0xff); out.write((v shr 8) and 0xff) }
            out.write("RIFF".toByteArray()); le32(totalSize); out.write("WAVE".toByteArray())
            out.write("fmt ".toByteArray()); le32(16); le16(1); le16(channels); le32(sampleRate); le32(byteRate); le16(channels * 2); le16(16)
            out.write("data".toByteArray()); le32(dataSize); out.write(pcm)
        }
    }

    private fun stopEverything() {
        if (!running && projection == null && audioRecord == null) {
            stopForeground(STOP_FOREGROUND_REMOVE)
            stopSelf()
            return
        }
        running = false
        lastStatus = "Đã dừng."
        try { audioRecord?.stop() } catch (_: Throwable) { }
        try { audioRecord?.release() } catch (_: Throwable) { }
        audioRecord = null
        try { projection?.stop() } catch (_: Throwable) { }
        projection = null
        latestChunk.getAndSet(null)?.delete()
        pendingSpeech.set(null)
        try { tts?.stop() } catch (_: Throwable) { }
        abandonDuck()
        translator?.close(); translator = null
        whisperModel?.let { try { Whisper.releaseModel(it) } catch (_: Throwable) { } }
        whisperModel = null
        scope.coroutineContext.cancelChildren()
        stopForeground(STOP_FOREGROUND_REMOVE)
        stopSelf()
    }

    override fun onDestroy() {
        running = false
        try { audioRecord?.release() } catch (_: Throwable) { }
        try { projection?.stop() } catch (_: Throwable) { }
        translator?.close()
        whisperModel?.let { try { Whisper.releaseModel(it) } catch (_: Throwable) { } }
        tts?.shutdown()
        scope.cancel()
        super.onDestroy()
    }

    private fun createChannel() {
        if (Build.VERSION.SDK_INT >= 26) {
            val channel = NotificationChannel(CHANNEL_ID, "LiveDub Việt", NotificationManager.IMPORTANCE_LOW).apply {
                description = "Giữ dịch vụ bắt âm thanh nội bộ hoạt động"
                setSound(null, null)
            }
            getSystemService(NotificationManager::class.java).createNotificationChannel(channel)
        }
    }
}
